controller/src/main/java/io/pravega/controller/store/stream/ZKStream.java

```diff
@@ -64,32 +76,29 @@
     private static final String STREAM_ACTIVE_TX_PATH = ZKStreamMetadataStore.ACTIVE_TX_ROOT_PATH + "/%s/%S";
     private static final String STREAM_COMPLETED_TX_BATCH_PATH = ZKStreamMetadataStore.COMPLETED_TX_BATCH_PATH + "/%s/%s";
 
-    private static final Data EMPTY_DATA = new Data(null, new Version.IntVersion(Integer.MIN_VALUE));
-
     private final ZKStoreHelper store;
-    private final String creationPath;
-    private final String configurationPath;
-    private final String truncationPath;
-    private final String statePath;
-    private final String epochTransitionPath;
-    private final String committingTxnsPath;
-    private final String waitingRequestProcessorPath;
-    private final String activeTxRoot;
-    private final String markerPath;
-    private final String idPath;
-    private final String scopePath;
-    @Getter(AccessLevel.PACKAGE)
-    private final String streamPath;
-    private final String retentionSetPath;
-    private final String retentionStreamCutRecordPathFormat;
-    private final String currentEpochRecordPath;
-    private final String epochRecordPathFormat;
-    private final String historyTimeSeriesChunkPathFormat;
-    private final String segmentSealedEpochPathFormat;
-    private final String segmentsSealedSizeMapShardPathFormat;
-
-    private final Cache cache;
+    private final AtomicReference<String> creationPath = new AtomicReference<>();
```

Comment:
This is changing the immutability of the state of objects of this class. Could we optimize without breaking immutability?

Reply:
Actually all these strings are created per Stream object irrespective of whether we want to access those paths or not. This was wasteful. 
Hence i loaded them in an atomic reference on the need basis. so now its loaded once and used. 
since this is an internal private field which is not exposed to anyone outside this class, so breaking the immutability for this field for the gains we get IMO is ok.

Reply:
You are right about them being private, but why do you want to lazily load these strings rather than loading them in the constructor like you are doing at the moment?

Reply:
its just a minor optimization to lazily load the evaluated strings on the need basis rather than load them at the time of construction of the object. most of the time we will only need to answer only a small subset of queries and would not need to have all the strings computed but only those that are needed by those queries.

Reply:
I don't understand the 'most of the time' part. Why would such an object be created over and over? Can we at least factor out this 'get__Path' logic into a helper or something that can be long lived. It's weird to intermix that with a data object.

Reply:
let me explain what we are doing here:
a stream object has multiple types of metadata records. each stored at a different path. 
most queries only deal with specific metadata records. 
For example writers or readers asking for successors. 
We were originally computing all metadata paths for a stream in the constructor of this Stream object. Which meant we were computing paths that may not even be required by queries coming in. 
It is important to note that controller doesnt keep these stream objects forever in its memory. 
They are created on demand, and kept in a cache for a brief period until they are regularly accessed. 
The controller will instantiate the object, and this object will be shared across multiple queries. Once no activity happens on the object for a while, it will be GC'd. 
So `most of the time` is about the kind of queries coming in during the lifespan of the object. 
If the queries coming in are typically `gets` that only require access to a small set of paths, then evaluating all the paths before hand may be wasteful. 

So for a query like getSuccessor we only need a handful of these paths to be evaluated. So evaluating and keeping all the strings in the object is wasteful. Each string being evaluated is about 50-100 chars or having a footprint of 100-200 bytes. The atomic reference object has a smaller footprint in comparison. So with 20 odd string paths, we maybe saving on maybe around 1000-2000 bytes per stream object. I dont know if the gains are significant enough to do this. 
but that is the reason for this choice. 

> Can we at least factor out this 'get__Path' logic into a helper 
So we can create a helper for computing the paths. but paths are specific to the stream in question.. so the helper cannot keep all the paths for all the streams.. and we dont want to compute the path every time before accessing the path. that would be more wasteful. 
so we want to keep the computed path in the stream object for reuse.

Reply:
On second thoughts, the gains here are not much and the complexity increase is more.. i will revert these changes.